src/backend/distributed/executor/executor.c

```diff
@@ -1844,14 +1902,19 @@ PlacementExecutionDone(TaskPlacementExecution *placementExecution, bool succeede
 		executionOrder == EXECUTION_ORDER_SEQUENTIAL)
 	{
 		TaskPlacementExecution *nextPlacementExecution = NULL;
-		int placementExecutionCount PG_USED_FOR_ASSERTS_ONLY =
-			shardCommandExecution->placementExecutionCount;
+		int placementExecutionCount = shardCommandExecution->placementExecutionCount;
 
 		/* find a placement execution that is not yet marked as failed */
 		do {
 			int nextPlacementExecutionIndex =
 				placementExecution->placementExecutionIndex + 1;
 
+			if (nextPlacementExecutionIndex == placementExecutionCount)
+			{
+				/* adjust the index if the placement execution is the last one */
+				nextPlacementExecutionIndex = 0;
```

Comment:
Don't we always start at 0 and count up (meaning we shouldn't have to wrap around)?

Line above this `if` statement is a bit suspicious seems `placementExecution doesn't change in this loop.

Reply:
As far as I understand, if the first placement execution fails, we might start from > 0?

Reply:
But when do we need to wrap around?

Reply:
That assumption that's encoded into this function is that we only reach this code if the task has not failed yet. That means that the placement execution that has failed cannot be the last one and therefore placementExecutionIndex + 1 cannot exceed the count.

However, this assumption is violated when a worker pool fails and we mark all placement executions on that worker as failed.

The `if` statement that checks the execution order originally had an additional condition which checked that the state of placementExecution was RUNNING at the start of the function, which means it must have been the "current" placement execution. 

In general, this is my least favourite part of the code because it's built on implicit assumptions and that is already leading to bugs. It could use some refactoring.

Reply:
Did you discuss this internally? I have similar concerns to Marco. Isn't that the case because we try the placements sequentially, we have already tried all indexes before `placementExecutionIndex`, so we don't need to wrap around?